<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>Legofit: $title</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">Legofit
   </div>
   <div id="projectbrief">Legofit infers population history from nucleotide site  patterns.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="contents">
<div class="textblock"><p>Generate site patterns by coalescent simulation</p>
<h1><code>legosim</code>: coalescent simulations within a network of populations</h1>
<pre class="fragment">usage: legosim [options] input_file
   where options may include:
   -i &lt;x&gt; or --nItr &lt;x&gt;
      number of iterations in simulation
   -1 or --singletons
      Use singleton site patterns
   -U &lt;x&gt;
      Mutations per generation per haploid genome.
   -h or --help
      print this message
</pre><p>Here, "input_file" should be in <a class="el" href="index.html#lgo">.lgo</a> format, which describes the history of population size, subdivision, and gene flow. By default (i.e. if the <code>-U</code> option is not used), the output looks like this: </p><pre class="fragment">############################################################
# legosim: generate site patterns by coalescent simulation #
############################################################

# Program was compiled: Dec 25 2016 10:10:51
# Program was run: Sun Dec 25 10:13:47 2016

# cmd: ./legosim -i 10000 input.lgo
# nreps                       : 10000
# input file                  : input.lgo
# not simulating mutations
# excluding singleton site patterns.
#       SitePat E[BranchLength]
            x:y      39.7970280
            x:n      38.9656878
            y:n      40.8560014
</pre><p>Here, the "SitePat" column labels site patterns. For example, site pattern xy (denoted by "x:y" in this output) refers to nocleotide sites at which the derived allele is present in single haploid samples from X and Y but not in samples from other populations. This site pattern arises when a mutation strikes a branch that is ancestral only to the samples from X and Y. The average length of this branch in generations appears under "E[BranchLength]".</p>
<p>To simulate site pattern counts across an entire genome, use the <code>-U</code> option, whose argument give the expected number of mutations per generation per haploid genome. This argument should equal <img class="formulaInl" alt="$\mu L$" src="form_0.png"/>, where <img class="formulaInl" alt="$\mu$" src="form_1.png"/> is the mutation rate per nucleotide site per generation, and <img class="formulaInl" alt="$L$" src="form_2.png"/> is the number of nucleotide sites sequenced per haploid genome, including monomorphic sites but excluding those that fail quality control. For example, adding <code>-U 18</code> to the command above led to the following output: </p><pre class="fragment">############################################################
# legosim: generate site patterns by coalescent simulation #
############################################################

# Program was compiled: Dec 25 2016 10:10:51
# Program was run: Mon Dec 26 09:31:59 2016

# cmd: ./legosim -i 10000 -U 18 input.lgo
# nreps                       : 10000
# input file                  : input.lgo
# mutations per haploid genome: 18.000000
# excluding singleton site patterns.
#       SitePat           Count
            x:y             770
            x:n             713
            y:n             677
</pre><p>Now, the 2nd column is labeled "Count" rather than "E[BranchLength]" and gives the simulated count of each site pattern across the genome as a whole. It is calculated by sampling from a Poisson distribution with mean <img class="formulaInl" alt="$\mu L b$" src="form_3.png"/>, where <img class="formulaInl" alt="$\mu$" src="form_1.png"/> and <img class="formulaInl" alt="$L$" src="form_2.png"/> are as described above, and <img class="formulaInl" alt="$b$" src="form_4.png"/> is the average branch length as reported without the <code>-U</code> option. This Poisson model treats nucleotide sites as independent, ignoring linkage disequilibrium. The counts it provides are correct in expectation, but their variances in repeated runs of the program are probably too small.</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2015, 2016, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span style="display: none;">.nosp@m.</span>s@an<span style="display: none;">.nosp@m.</span>thro.<span style="display: none;">.nosp@m.</span>utah<span style="display: none;">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Tue Aug 11 2020 15:55:25 for Legofit by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
