<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.13"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>legofit: $title</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">legofit
   </div>
   <div id="projectbrief">Legofit is a system of programs written in C and Python,whichusescountsofnucleotidesitepatternstoinferthe</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.13 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="contents">
<div class="textblock"><p>Estimate parameters describing population sizes, the times of separations and of episodes of gene flow, and levels of gene flow.</p>
<h1><code>legofit</code>: estimate population history from site pattern data</h1>
<pre class="fragment">usage: legofit [options] input.lgo sitepat.txt
   where file input.lgo describes population history,
   and file sitepat.txt contains site pattern frequencies.
Options may include:
   -T &lt;x&gt; or --tol &lt;x&gt;
      termination criterion
   -t &lt;x&gt; or --threads &lt;x&gt;
      number of threads (default is auto)
   -F &lt;x&gt; or --scaleFactor &lt;x&gt;
      set DE scale factor
   -x &lt;x&gt; or --crossover &lt;x&gt;
      set DE crossover probability
   -s &lt;x&gt; or --strategy &lt;x&gt;
      set DE strategy
   -S &lt;g&gt;@&lt;r&gt; or --stage &lt;g&gt;@&lt;r&gt;
      add stage with &lt;g&gt; generations and &lt;r&gt; simulation reps
   -p &lt;x&gt; or --ptsPerDim &lt;x&gt;
      number of DE points per free var
   --stateIn &lt;filename&gt;
      read initial state of optimizer from file. Option may be repeated.
   --stateOut &lt;filename&gt;
      write final state of optimizer to file
   -1 or --singletons
      Use singleton site patterns
   -v or --verbose
      verbose output
   --version
      Print version and exit
   -h or --help
      print this message
</pre><p>Two arguments are required:</p>
<ul>
<li>an input file in <a class="el" href="index.html#lgo">.lgo</a> format, which describes the history of population size, subdivision, and gene flow;</li>
<li>a file in the format produced by <a class="el" href="tabpat.html">tabpat</a>, which provides counts of site patterns in the data.</li>
</ul>
<p>Legofit estimates the values of all the parameters that are declared as "free" in the .lgo file. It does this by minimizing a "cost
function", which measures the difference between observed and expected values. Currently, the cost function is the negative of composite likelihood. Other options are available via compile-time option (see <code><a class="el" href="typedefs_8h_source.html">typedefs.h</a></code>). I don't yet know which cost function is best.</p>
<p>Expected counts are estimated by computer simulation, and optimization is done using the "differential evolution" (DE) algorithm. The DE algorithm maintains a swarm of points, each at a different set of parameter values. The objective function is evaluated at these points in a multithreaded job queue, so the program runs fastest on a machine with lots of cores. You can set the number of threads using the <code>-t</code> argument. By default, the program uses 3/4 as many threads as there are processors on the machine&mdash;usually the number of hypercores.</p>
<p>The DE algorithm can be tuned via command line arguments <code>-F</code>, <code>-x</code>, <code>-s</code>, and <code>-p</code>. Details regarding these choices can be found in "Differential evolution: a practical approach to global optimization", by Price, Storn, and Lampinen. We've had good results with <code>-s 2</code>, <code>-F 0.3</code>, and <code>-x 0.8</code>, so these became the defaults as of version 1.15.</p>
<p>During the first few hundred generations of the DE algorithm, the swarm of points adapts to the objective function. During this initial phase, high accuracy is not required, so it can speed things up to use low resolution in function evaluations. This is the purpose of the <code>-S</code> argument, which adds a "stage" to the simulation schedule. For example, <code>-S 1000@10000</code> adds a stage of 1000 DE generations, in each of which 10000 simulation replicates are used to evaluate each objective function. The <code>-S</code> argument can be given several times to set up a simulation schedule with several stages. The algorithm is allowed to converge only during the final stage. I am currently using a 2-stage schedule: <code>-S 1000@10000 -S</code>1000@2000000`.</p>
<p>By default, the initial swarm of points consists of one point representing the parameter values in the .lgo file, plus other points scattered randomly throughout the feasible region of parameter space. The total number of points defaults to 10 times the number of free parameters. To change this number, see the &ndash;ptsPerDim option.</p>
<p>The initial swarm of points can also be specified using the <code>--stateIn</code> option. This reads a file specifying the initial state of the swarm of points maintained by DE. The format of this file is as described below for the <code>--stateOut</code> option. The number of free parameters in each <code>--stateIn</code> file should be as specified in the .lgo file. The number of points in the files may differ.</p>
<p>The <code>--stateIn</code> option may be given more than once, each time with a different input file. When more than one file is given, Legofit constructs the initial swarm of points by combining points from all input files.</p>
<p>The option <code>--stateOut</code> is used to define an output file for the final state of the optimizer. This output file begins with a row giving the number of points and the number of free parameters. After that, there is a row for each point in the swarm of points maintained by <a class="el" href="diffev_8c.html">diffev.c</a>. In each row, the first entry is the value of the cost function at that point. The remaining entries give the free parameter values in the same order in which they are printed by legofit.</p>
<p>The format of the state file changed in early July, 2018. Before that date, the state file did not include names of parameters. Parameter values in the state file had to be arranged in the same order as the free parameters in the .lgo file. If the .lgo and state files referred to different parameters or to the same parameters in a different order, parameters would get the wrong values. No error was detected unless this misassignment resulted in a tree that was not feasible&ndash;for example, one in which a segment of the population tree was older than its parent in the tree.</p>
<p>The new state file format includes the names of parameters, and the input routine compares these against the names of free parameters in the .lgo file. The two lists must have the same parameters in the same order. Otherwise, legofit aborts with an error message. Old- and new-format state files can both be input using &ndash;stateIn arguments and can be intermingled in a single legofit run.</p>
<p>The <code>-1</code> option tells legofit to use singleton site patterns&ndash;patterns in which the derived allele is present in only a single sample. This is a bad idea with low-coverage sequence data, because singletons are likely to be sequencing errors. We use this option routinely now with high-coverage data. To use it, you need to generate a data set that contains singletons, by using the <code>-1</code> option of \ref tabpat "tabpat", <a class="el" href="sitepat.html">sitepat</a>, <a class="el" href="scrmpat.html">scrmpat</a>, or <a class="el" href="simpat.html">simpat</a>.</p>
<p>2017-08-26: The convergence criterion has been changed. In the previous code, differential evolution (DE) iterations stopped when neither the best objective function value nor the spread of these values had changed in a fixed number of iterations. That criterion was used in our PNAS paper, "Early history of Neanderthals and
Denisovans".</p>
<p>I began to notice convergence problems with models larger than those used in the August 2017 PNAS paper. All bootstrap replicates would report convergence, but some yielded wild parameter estimates. In these outliers, the spread of objective function values was also very large, indicating that the algorithm had not really converged. So I implemented a new convergence criterion, based on the spread of objective function values. The iterations terminate when this spread falls to a pre-determined value.</p>
<p>This new convergence criterion works best with the KL (Kullback-Leibler) cost function. Minimizing KL is the same as maximizing log composite likelihood (lnL). But KL equals zero when the fit is perfect. For this reason, it isn't necessary to scale the tolerance value to lnL.</p>
<p>In each DE generation, the algorithm evaluates the cost function of each point in the swarm. The cost function measures the difference between observed and expected site pattern frequencies. The difference between the maximum and minimum cost is called the "spread". The algorithm stops when "spread" falls to the tolerance value, "ytol". This tolerance can be set on the command line with "-T" or "--tol" arguments. Convergence is checked only in the last stage, as set by "-S" or "--stage".</p>
<p>If the algorithm fails to converge, there are several options. First, you can use "-S" or "--stage" to increase the maximum number of DE generations. For example, if the last stage is "-S 2000@2000000", you are allowing 2000 DE generations, in each of which each cost function value is estimated with 2000000 iterations. To double the allowed number of DE generations, change this to "-S 4000@2000000".</p>
<p>Second, you can relax the tolerance. By default, this is 1e-4. It is reported in the legofit output. To double this value, use "-T 2e-4" or "--tol 2e-4".</p>
<p>Legofit handles three types of signal: SIGINT, SIGTERM, and SIGUSR1. If legofit is running in the foreground, the first of these signals can be generated by typing Ctrl-C. Otherwise (under linux or osx), use <code>killall -SIGINT legofit</code>, <code>killall -SIGTERM legofit</code>, or <code>killall -SIGUSR1 legofit</code>. In response to SIGINT or SIGTERM, Legofit will wait until the end of the current generation of the diffev algorithm and then exit gracefully, printing all the usual output. In response to SIGUSR1, Legofit will wait until the end of the current DE generation and then write to stderr a summary of the state of the optimizer.</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2016, 2017, 2018, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span style="display: none;">.nosp@m.</span>s@an<span style="display: none;">.nosp@m.</span>thro.<span style="display: none;">.nosp@m.</span>utah<span style="display: none;">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Tue Aug 11 2020 15:40:03 for legofit by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.13
</small></address>
</body>
</html>
