<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.11"/>
<title>lego: $title</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { init_search(); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">lego
   &#160;<span id="projectnumber">0.1</span>
   </div>
   <div id="projectbrief">Lego is a system of programs written in C and Python, which uses counts of nucleotide site patterns to infer thehistoryofpopulationsize,subdivision,andgeneflow.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.11 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li class="current"><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="contents">
<div class="textblock"><p>Estimate parameters describing population sizes, the times of separations and of episodes of gene flow, and levels of gene flow.</p>
<h1><code>legofit</code>: estimate population history from site pattern data</h1>
<pre class="fragment">usage: legofit [options] -u &lt;mut_rate&gt; -n &lt;genome_size&gt; input.lgo sitepat.txt
   where &lt;mut_rate&gt; is the mutation rate per nucleotide
   site per generation, &lt;genome_size&gt; is the number of
   nucleotides per haploid genome, file input.lgo describes
   population history, and file sitepat.txt contains site
   pattern frequencies.
Options may include:
   -M &lt;x&gt; or --maxFlat &lt;x&gt;
      termination criterion
   -t &lt;x&gt; or --threads &lt;x&gt;
      number of threads (default is auto)
   -F &lt;x&gt; or --scaleFactor &lt;x&gt;
      set DE scale factor
   -x &lt;x&gt; or --crossover &lt;x&gt;
      set DE crossover probability
   -s &lt;x&gt; or --strategy &lt;x&gt;
      set DE strategy
   -S &lt;g&gt;@&lt;r&gt; or --stage &lt;g&gt;@&lt;r&gt;
      add stage with &lt;g&gt; generations and &lt;r&gt; simulation reps
   -p &lt;x&gt; or --ptsPerDim &lt;x&gt;
      number of DE points per free var
   -1 or --singletons
      Use singleton site patterns
   -v or --verbose
      verbose output
   -h or --help
      print this message
</pre><p>Four arguments are required:</p>
<ul>
<li><code>-u</code>, the mutation rate per nucleotide site per generation;</li>
<li><code>-n</code>, the number of nucleotide sites sequenced, including monomorphic sites but not including sites that failed quality control criteria;</li>
<li>an input file in <a class="el" href="index.html#lgo">.lgo</a> format, which describes the history of population size, subdivision, and gene flow;</li>
<li>a file in the format produced by <a class="el" href="tabpat.html">tabpat</a>, which provides counts of site patterns in the data.</li>
</ul>
<p>Legofit estimates the values of all the parameters that are declared as "free" (rather than "fixed" or "gaussian") in the .lgo file. It does this by minimizing a "cost function", which measures the difference between observed and expected values. Currently, the cost function is a Chi-square statistic. Other options are available via compile-time option (see <code><a class="el" href="typedefs_8h_source.html">typedefs.h</a></code>). One of the alternatives, KL_COST, would not require the -u and -n command-line arguments. On the other hand, it would not provide an internal molecular clock&ndash;the clock would be calibrated only by the time parameters in the .lgo file. I don't yet know which cost function is best.</p>
<p>Expected counts are estimated by computer simulation, and optimization is done using the "differential evolution" (DE) algorithm. The DE algorithm maintains a swarm of points, each at a different set of parameter values. The objective function is evaluated at these points in a multithreaded job queue, so the program runs fasted on a machine with lots of cores. You can set the number of threads using the <code>-t</code> argument. By default, the program uses as many threads as there are processors on the machine&mdash;usually the number of hypercores. The optimal number of threads is usually somewhat smaller that this default.</p>
<p>The DE algorithm can be tuned via command line arguments <code>-F</code>, <code>-x</code>, <code>-s</code>, and <code>-p</code>. Details regarding these choices can be found in "Differential evolution: a practical approach to global optimization", by Price, Storn, and Lampinen. I don't yet know what is best, but I'm currently using <code>-s 2</code> and <code>-s 4</code>, with default values of the other options.</p>
<p>During the first few hundred generations of the DE algorithm, the swarm of points adapts to the objective function. During this initial phase, high accuracy is not required, so it can speed things up to use low resolution in function evaluations. This is the purpose of the <code>-S</code> argument, which adds a "stage" to the simulation schedule. For example, <code>-S 1000@10000</code> adds a stage of 1000 DE generations, in each of which 10000 simulation replicates are used to evaluate each objective function. The <code>-S</code> argument can be given several times to set up a simulation schedule with several stages. The algorithm is allowed to converge only during the final stage. I am currently using a 2-stage schedule: <code>-S 1000@10000 -S</code>1000@2000000`.</p>
<p>The <code>-1</code> option tells legofit to use singleton site patterns&ndash;patterns in which the derived allele is present in only a single sample. This is a bad idea with low-coverage sequence data. It also behaves poorly with archaic DNA, because singletons are likely to be sequencing artifacts. With high-quality modern DNA, it may one day prove useful. To use it, you need to generate a data set that contains singletons, by using the <code>-1</code> option of \ref tabpat "tabpat".</p>
<dl class="section copyright"><dt>Copyright</dt><dd>Copyright (c) 2016, Alan R. Rogers <a href="#" onclick="location.href='mai'+'lto:'+'rog'+'er'+'s@a'+'nt'+'hro'+'.u'+'tah'+'.e'+'du'; return false;">roger<span style="display: none;">.nosp@m.</span>s@an<span style="display: none;">.nosp@m.</span>thro.<span style="display: none;">.nosp@m.</span>utah<span style="display: none;">.nosp@m.</span>.edu</a>. This file is released under the Internet Systems Consortium License, which can be found in file "LICENSE". </dd></dl>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Sat Dec 31 2016 17:48:09 for lego by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.11
</small></address>
</body>
</html>
