/**
@file msppat.c
@page msppat
@brief Tabulate site pattern frequencies from .daf files.

# Scrmpat: tabulates site patterns

Scrmpat reads data generated by msprime, tabulates counts of
nucleotide site patterns, writing the result to standard
output.

# Usage

    Usage: msppat [options]
      where options may include:
       --version
          Print version and exit
       -h or --help
          Print this message

# Example

`msppat` parses a file generated using `msprime`. Let us assume you
have done this, that file `foo.msp` contains the output simulated
by `msprime`, and that these simulated data included genotypes referring
to four populations, labeled "x", "y", "n", and "d". The `msppat`
command would look like this:

    msppat foo.msprime x y n d

`msppat`'s notion of a "population" differs from that of `msprime`, in
that `msppat` treats samples of different ages as separate
populations, even if they reside in the same population on the `msprime`
command line. For example, consider the following `msprime` command line:

    msprime 3 -I 2 1 1 -eI 0.5 0 1

This specifies three haploid samples distributed across two
populations. The `-I` argument says that each population has a
sample at time 0. The `-eI` argument says that, in addition,
population 2 has a sample at time 0.5. All three samples would be
treated as separate populations by `msppat`. Thus, the `msppat`
command line should list three labels, as in "msppat x y z".

In the output, site pattern "x:y" refers to
the pattern in which the derived allele is present haploid samples
from "x" and "y" but not on those from other populations. The order of
the command-line arguments determines the order in which labels are
sorted on output. Given the command line above, we would get a site
pattern labeled "x:y:d" rather than, say, "y:x:d".

The output looks like this:

    # msppat version 1.3
    # Population labels: x y n d
    # Number of site patterns: 10
    # Tabulated 12327755 SNPs
    #       SitePat             E[count]
                x:y       340952.4592501
                x:n        46874.1307236
                x:d        46034.4670204
                y:n        55137.4236715
                y:d        43535.5248078
                n:d       231953.3372578
              x:y:n        91646.1277991
              x:y:d        88476.9619569
              x:n:d        96676.3877423
              y:n:d       100311.4411513

The left column lists the site patterns that occur in the data. The
right column gives the expected count of each site pattern. These are
not integers, because they represent averages over all possible
subsamples consisting of a single haploid genome from each
population.

To generate a bootstrap, use the `--bootreps` option:

    msppat --bootreps 50 --infile foo.msprime x y n d > obs.txt

This will generate not only the primary output file, `obs.txt`, but also
50 additional files, each representing a single bootstrap
replicate. The primary output file now has a bootstrap confidence
interval:

    # Including singleton site patterns.
    # Number of site patterns: 10
    # Tabulated 12327755 SNPs
    # bootstrap output file = msppatX.bsr
    # confidence level = 95%
    #       SitePat             E[count]          loBnd          hiBnd
                x:y       340952.4592501 338825.6604586 342406.6670816
                x:n        46874.1307236  46361.5798377  47438.1857029
                x:d        46034.4670204  45605.6588012  46631.6434277
                y:n        55137.4236715  54650.0763578  55783.7051253
                y:d        43535.5248078  43110.5119922  44234.0919024
                n:d       231953.3372578 229495.3741057 234173.6878092
              x:y:n        91646.1277991  90494.0219749  92873.4443706
              x:y:d        88476.9619569  87137.1867967  89585.8431419
              x:n:d        96676.3877423  95935.5184294  97417.6241185
              y:n:d       100311.4411513  99292.9839140 101163.3457462

Here, `loBnd` and `hiBnd` are the limits of a 95% confidence
interval. The bootstrap output files look like `msppat0.bsr`,
`msppat1.bsr`, and so on.

@copyright Copyright (c) 2018, Alan R. Rogers
<rogers@anthro.utah.edu>. This file is released under the Internet
Systems Consortium License, which can be found in file "LICENSE".
*/

#include "binary.h"
#include "misc.h"
#include "mspreader.h"
#include "typedefs.h"
#include "version.h"
#include "error.h"
#include <ctype.h>
#include <errno.h>
#include <getopt.h>
#include <gsl/gsl_rng.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef struct Stack Stack;

/// Treat a vector of tipId_t values as a push-down stack.
struct Stack {
    int         dim, nused;
    tipId_t    *buff;           // not locally owned
};

static void usage(void);
static Stack *Stack_new(int dim, tipId_t buff[dim]);
static void Stack_free(Stack * stk);
static void Stack_push(Stack * self, tipId_t x);
static void generatePatterns(int bit, int npops, Stack * stk, tipId_t pat,
                             int doSing);

const char *useMsg =
    "\nUsage: msppat [options] [inputfile]\n"
    "  Reads from stdin by default. Options may include:\n";

/// Print usage message and die.
static void usage(void) {
    fputs(useMsg, stderr);
    tellopt("-V or --version", "Print version and exit");
    tellopt("-h or --help", "Print this message");
    exit(1);
}

/// This stack is local to this file. It provides a bounds-controlled
/// interface to an external array, which is passed as an argument, buff,
/// to Stack_new.
static Stack *Stack_new(int dim, tipId_t buff[dim]) {
    Stack      *self = malloc(sizeof(Stack));
    CHECKMEM(self);
    self->dim = dim;
    self->buff = buff;
    self->nused = 0;
    return self;
}

/// Frees the stack but not the underlying buffer.
static void Stack_free(Stack * stk) {
    free(stk);
}

/// Add an entry to the tail of the stack, checking bounds.
static void Stack_push(Stack * self, tipId_t x) {
    if(self->nused == self->dim) {
        fprintf(stderr, "%s:%s:%d ERR: buffer overflow\n",
                __FILE__, __func__, __LINE__);
        exit(EXIT_FAILURE);
    }
    self->buff[self->nused++] = x;
}

/// Call as generatePatterns(0, npops, stk, 0); Recursive function,
/// which generates all legal site patterns and pushes them onto a
/// stack.
static void
generatePatterns(int bit, int npops, Stack * stk, tipId_t pat, int doSing) {
    assert(sizeof(tipId_t) < sizeof(unsigned long long));
    if(bit == npops) {
        // Recursion stops here. If current pattern is
        // legal, then push it onto the stack. Then return.

        // Exclude patterns with all bits on, or all bits off.
        if(pat == 0 || pat == (1ULL << npops) - 1ULL)
            return;
        // Exclude singleton patterns unless "doSing" is true.
        if(!doSing && isPow2(pat))
            return;
        Stack_push(stk, pat);
        return;
    }
    tipId_t     on = 1UL << bit;
    generatePatterns(bit + 1, npops, stk, pat | on, doSing);    // curr bit on
    generatePatterns(bit + 1, npops, stk, pat, doSing); // curr bit off
}

int main(int argc, char **argv) {
    int         i, j, status, optndx, done;
    int         doSing = 1;     // nonzero means use singleton site patterns
    char        errbuff[100] = { '\0' };
    FILE       *ifp = stdin;

    static struct option myopts[] = {
        // {char *name, int has_arg, int *flag, int val}
        {"help", no_argument, 0, 'h'},
        {"version", no_argument, 0, 'V'},
        {NULL, 0, NULL, 0}
    };

    // command line arguments
    for(;;) {
        i = getopt_long(argc, argv, "hV", myopts, &optndx);
        if(i == -1)
            break;
        switch (i) {
        case '?':
            usage();
            break;
        case 'h':
            usage();
            break;
        case 'V':
            printf("msppat version %s\n", VERSION);
            return 0;
        default:
            usage();
        }
    }

    // There can be one additional argument, the name of the input
    // file.
    switch(argc - optind) {
    case 0:
        ifp = stdin;
        break;
    case 1:
        const char *fname = argv[optind + 1];
        ifp = fopen(fname, "r");
        if(ifp==NULL) {
            fprintf(stderr,"%s:%d: can't open %s for input.\n",
                    __FILE__,__LINE__, fname);
            exit(EXIT_FAILURE);
        }
        break;
    default:
        fprintf(stderr,"Only one input file is allowed.\n");
        usage();
    }

    MsprimeReader *r=MsprimeReader_new(ifp);
    if(r == NULL) {
        fprintf(stderr,"%s:%d: Can't read input\n",
                __FILE__,__LINE__);
        exit(EXIT_FAILURE);
    }
    int n = MsprimeReader_sampleDim(r);

    LblNdx lblndx;
    LblNdx_init(&lblndx);
    for(i=0; i < MsprimeReader_sampleDim(r); ++i) {
        LblNdx_addSamples(&lblndx,
                          MsprimeReader_nsamples(r, i),
                          MsprimeReader_lbl(r, i))
    }

    printf("# msppat version %s\n", VERSION);

    unsigned long npat = (1UL << n) - 2UL;  // number of site patterns
    if(!doSing)
        npat -= n;
    printf("# %s singleton site patterns.\n",
           (doSing ? "Including" : "Excluding"));
    printf("# Number of site patterns: %lu\n", npat);
    tipId_t     pat[npat];
    double      patCount[npat];
    int         lblsize = 100;
    char        lblbuff[lblsize];
    memset(patCount, 0, sizeof(patCount));

    {
        // Stack is a interface to array "pat".
        Stack      *stk = Stack_new(npat, pat);

        // Put site patterns into array "pat".
        generatePatterns(0, n, stk, 0, doSing);

        Stack_free(stk);
    }

    // Sort site patterns. Major sort is by number of "on" bits,
    // so that singleton patterns come first, then doubletons, etc.
    // Secondary sort is by order in which labels are listed
    // on the command line.
    qsort(pat, (size_t) npat, sizeof(pat[0]), compare_tipId);
    fflush(stdout);

    unsigned long nsites = 0, nbadaa = 0, nfixed = 0;
    long snpndx = -1;
    int chrndx=0;

    // Read data
    fprintf(stderr, "Doing %s pass through data to tabulate patterns..\n",
            "single");
    done=0;
    while(!done) {
        status = MsprimeReader_next(r);
        switch(status) {
        case 0:
            ++nsites;
            break;
        case EOF:
            done=1;
            continue;
        case ALLELE_MISMATCH:
        case NO_ANCESTRAL_ALLELE:
            ++nbadaa;
            ++nsites;
            continue;
        default:
            // something wrong
            mystrerror_r(status, errbuff, sizeof errbuff);
            fprintf(stderr,"%s:%d: input error (%s)\n",
                    __FILE__,__LINE__, errbuff);
            exit(EXIT_FAILURE);
        }

        ++snpndx;

        // p and q are frequencies of derived and ancestral alleles
        double p[n], q[n];
        for(j = 0; j < n; ++j) {
            p[j] = MsprimeReader_daf(r,j); // derived allele freq
            q[j] = 1.0 - p[j];
        }

        // Contribution of current snp to each site pattern.  Inner
        // loop considers each bit in current pattern.  If that bit is
        // on, multiply z by the derived allele frequency, p. If
        // that bit is off, multiply by q=1-p. In the end, z is Prod
        // p[j]^bit[j] * q[j]^(1-bit[j]) where bit[j] is the value (0
        // or 1) of the j'th bit.
        for(i = 0; i < npat; ++i) {
            tipId_t     pattern = pat[i];
            double      z = 1.0;
            for(j = 0; j < n; ++j) {
                if(pattern & 1u)
                    z *= p[j];
                else
                    z *= q[j];
                pattern >>= 1u;
            }
            if(!isfinite(z)) {
                fprintf(stderr, "%s:%d nonfinite z=%lf\n",
                        __FILE__, __LINE__, z);
                fprintf(stderr, "   pattern=%d\n", pat[i]);
                for(j = 0; j < n; ++j)
                    fprintf(stderr, "   %d: p=%lf q=%lf\n", j, p[j], q[j]);
            }
            assert(0 == (pattern & 1));
            patCount[i] += z;
        }
    }
    printf("# Nucleotide sites: %lu\n", nsites);
    if(nbadaa)
        printf("# Disagreements about alleles: %lu\n", nbadaa);
    if(nfixed)
        printf("# Monomorphic sites: %lu\n", nfixed);
    printf("# Sites used: %lu\n",
           nsites - nbadaa - nfixed);

    // print labels and site patterns
    printf("# %13s %20s", "SitePat", "E[count]");
    putchar('\n');
    for(i = 0; i < npat; ++i) {
        printf("%15s %20.7lf\n",
               patLbl(lblsize, lblbuff, pat[i], &lndx), patCount[i]);
    }

    MsprimeReader_free(r);
    if(ifp!=stdin)
        fclose(ifp);
    fprintf(stderr, "msppat is finished\n");
    return 0;
}
